packages/builder-vite/source-loader-plugin.ts

```diff
@@ -1,21 +1,88 @@
 import sourceLoaderTransform from '@storybook/source-loader';
+import type { ExtendedOptions } from './types';
 
-export function sourceLoaderPlugin() {
-  return {
-    name: 'storybook-vite-source-loader-plugin',
-    enforce: 'pre',
-    async transform(src: string, id: string) {
-      if (id.match(/\.stories\.[jt]sx?$/)) {
-        // We need to mock 'this' when calling transform from @storybook/source-loader
-        // noinspection JSUnusedGlobalSymbols
-        const mockClassLoader = { emitWarning: (message: string) => console.warn(message), resourcePath: id };
-        const code = await sourceLoaderTransform.call(mockClassLoader, src);
-
-        return {
-          code,
-          map: { mappings: '' },
-        };
-      }
+const storyPattern = /\.stories\.[jt]sx?$/;
+const storySourcePattern = /var __STORY__ = "(.*)"/;
+const storySourceReplacement = '--STORY_SOURCE_REPLACEMENT--';
+
+const mockClassLoader = (id: string) => ({ emitWarning: (message: string) => console.warn(message), resourcePath: id });
+
+// HACK: Until we can support only node 15+ and use string.prototype.replaceAll
+const replaceAll = (str: string, search: string, replacement: string) => {
+  return str.split(search).join(replacement);
+};
+
+export function sourceLoaderPlugin(config: ExtendedOptions) {
+  if (config.configType === 'DEVELOPMENT') {
+    return {
+      name: 'storybook-vite-source-loader-plugin',
+      enforce: 'pre',
+      async transform(src: string, id: string) {
+        if (id.match(storyPattern)) {
+          const code = await sourceLoaderTransform.call(mockClassLoader(id), src);
+
+          return {
+            code,
+            map: { mappings: '' },
+          };
+        }
+      },
+    };
+  }
+
+  // In production, we need to be fancier, to avoid vite:define plugin from replacing values inside the `__STORY__` string
+  const storySources = new WeakMap<ExtendedOptions, Map<string, string>>();
+
+  return [
+    {
+      name: 'storybook-vite-source-loader-plugin',
+      enforce: 'pre',
+      buildStart() {
+        storySources.set(config, new Map());
+      },
+      async transform(src: string, id: string) {
+        if (id.match(storyPattern)) {
+          const code = await sourceLoaderTransform.call(mockClassLoader(id), src);
```

Comment:
There are some hidden type issues but nothing that breaks functionality.

Because this returns any, sourceString below is any even though the type is strictly string | null which will upset replaceAll but the any means this isn't thrown.